### DISCUSSION TITLE
How handle drag & drop between 2 list box ?

### DISCUSSION THREAD

**Discussion by Loulou007** [vdelachaux/UI-with-Classes#4] [Q&A] [answered]
Hi Vincent,

do you have a sample how manage d&d between listboxes? i.e. how refuse a drop during drop over (aka $0:=-1 in the old way)

Regards

**Comment by vdelachaux** [comment]
Hello,
No demo example yet, but here's what I do:
- On begin drag event, I place data with a unique signature in the pasteboard container. This may be a scalar value or, more usefully, an object that is stringified or embedded in a blob.
- When moving the target (On drag event), I test whether the data signature exists, if so, I accept it ($0:=0), if not, I reject it ($0:=-1). There's no other way to tell 4D what to do.
- If we accept the drag and the user actually drops the item on the target, we get the data from the pasteboard container and continue.

One important thing I'm discovering that you need to know is that it must be the object's method that returns 0 or -1. It can't be the class. That's the purpose of the “dropableWidgetMethod”.

And since you ask, I'll do a demo as soon as possible.

Regards
Vincent

**Comment by Loulou007** [comment]
Hi Vincent

thank you for the information. 

Regards

**Comment by Loulou007** [comment]
I somebody is looking for:
In script from target list box:



```4d
#DECLARE()->$result : Integer

var $o : Object
$o:=New object("f"; Formula(dropableWidgetMethod))
$result:=$o.f()
``` 
And in your class you need a function
ie name from your listbox is 'listbox_target'
```4d
Function _listbox_targetManager($e : cs.ui.evt)->$result : Integer
	$result:=0
	var $data_bl : Blob
	var $data_o : Object
	var $c : Collection
	Case of 
		: ($e.code=On Drag Over)
			GET PASTEBOARD DATA("appmaker.action.source"; $data_bl)
		        If (BLOB size($data_bl)>0)
			        $data:=BLOB to text($data_bl; UTF8 text without length)
			        
			        $data_o:=JSON Parse($data; Is object)
			       // Accept or reject the drag event 
			        If (($data_o.value%2)=0)
				        $result:=0
			        Else 
				        $result:=-1
			        End if 
		        End if 
			
		: ($e.code=On Drop)
			GET PASTEBOARD DATA("appmaker.action.source"; $data_bl)
			If (BLOB size($data_bl)>0)
				$data:=BLOB to text($data_bl; UTF8 text without length)
				
				$data_o:=JSON Parse($data; Is object)
				$c:=This.listbox_target.source
				$c.push($data_o)
				
				This.listbox_target.setSource($c)
				
				
			End if 
	End case 
```

**Comment by vdelachaux** [comment, marked as answer]
In fact, this mechanism is already implemented in a bit more generic way in the component: You have a shared method called “[dropableWidgetMethod](dropableWidgetMethod.4dm)”.

* Simply call this method in the listbox's object method.

> ```4d
> #DECLARE():Integer
> return dropableWidgetMethod
> ```

* Create a function in the dialog class called `“ _<form object name>Manager”`

The mechanism is equivalent to the one you have proposed.

There are also other shared methods, for subforms for example...

I really have to provide HDIs for these very specific cases.
